nixos/modules/services/web-servers/agate.nix

```
@@ -0,0 +1,122 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.agate;
+in
+{
+  options = {
+    services.agate = {
+      enable = mkEnableOption "Agate Server";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.agate;
+        defaultText = literalExpression "pkgs.agate";
+        description = "The package to use";
+      };
+
+      addresses = mkOption {
+        type = types.listOf types.str;
+        default = [ "0.0.0.0:1965" ];
+        description = ''
+          Addresses to listen on, IP:PORT, if you haven't disabled forwarding
+          only set IPv4.
+        '';
+      };
+
+      contentDir = mkOption {
+        default = "/var/lib/agate/content";
+        type = types.path;
+        description = "Root of the content directory.";
+      };
+
+      certificatesDir = mkOption {
+        default = [ "/var/lib/agate/certificates" ];
+        type = types.listOf types.path;
+        description = "Root of the certificate directory.";
+      };
+
+      hostnames = mkOption {
+        default = [ ];
+        type = types.listOf types.str;
+        description = ''
+          Domain name of this Gemini server, enables checking hostname and port
+          in requests. (multiple occurences means basic vhosts)
+        '';
+      };
+
+      language = mkOption {
+        default = null;
+        type = types.nullOr types.str;
+        description = "RFC 4646 Language code for text/gemini documents.";
+      };
+
+      onlyTls_1_3 = mkOption {
+        default = false;
+        type = types.bool;
+        description = "Only use TLSv1.3 (default also allows TLSv1.2).";
+      };
+
+      extraArgs = mkOption {
+        type = types.str;
+        default = "";
+        example = "--log-ip";
+        description = "Extra arguments to use running agate.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.agate =
+      let
+        convertListArgs = key: list: concatMapStringsSep " " (value: "${key} '${value}'") list;
+        addresses = convertListArgs "--addr" cfg.addresses;
+        hostnames = convertListArgs "--hostname" cfg.hostnames;
+      in
+      {
+        description = "Agate";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        preStart = ''
+          if [ ! -d '${cfg.contentDir}' ]; then
```

```suggestion
          if [ ! -d ${escapeShellArg cfg.contentDir} ]; then
```
also for all other occurences of straight interpolation.

both this and the following creation would be better expressed as `systemd.tmpfiles` rules as well, as it stands the script would be executed with user permissions and probably won't work reliably

---

I've gone for `StateDirectory` which seems to work well